<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>R packages and how to use them</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">[BETA] R tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">About</a>
</li>
<li>
  <a href="setup.html">setup</a>
</li>
<li>
  <a href="session1.html">session 1</a>
</li>
<li>
  <a href="packages.html">packages</a>
</li>
<li>
  <a href="session2.html">session 2</a>
</li>
<li>
  <a href="session3.html">session 3</a>
</li>
<li>
  <a href="reading.html">reading &amp; resources</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">R packages and how to use them</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#installing-packages">Installing packages</a></li>
<li><a href="#using-packages">Using packages</a>
<ul>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#citing-packages">citing packages</a></li>
</ul></li>
<li><a href="#key-packages">key packages</a>
<ul>
<li><a href="#tidyverse">Tidyverse</a></li>
<li><a href="#ggplot2">ggplot2</a></li>
<li><a href="#bioconductor">Bioconductor</a></li>
<li><a href="#r-markdown">R markdown</a></li>
</ul></li>
<li><a href="#namespaces">Namespaces</a></li>
<li><a href="#packages-for-session-2">Packages for session 2</a></li>
</ul>
</div>

<p>So far we’ve only used functions supplied as part of base R. However the real power of R comes from the ability to import functions the community of R programmers has built. These come in packages, which are a series of related functions bundled together</p>
<p>Generally speaking before implementing your function it is always worth checking if there is a well-maintained and documented package that performs what you would like to do.</p>
<div id="installing-packages" class="section level1">
<h1>Installing packages</h1>
<p>Most R packages belong to a repository known as CRAN. CRAN packages can be installed with a function</p>
<pre class="r"><code>install.packages(&quot;package_name&quot;)</code></pre>
<p>Often R will ask for your permission to install packages compiled from source rather than binary releases <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. If you want to use the absolutely most recent version say yes, if you’re in a hurry say no. Installing packages from source on Windows requires an additional component: <a href="https://cran.r-project.org/bin/windows/Rtools/">RTools</a>.</p>
</div>
<div id="using-packages" class="section level1">
<h1>Using packages</h1>
<p>packages can be loaded by using the library function. Conventionally, packages are loaded at the beginning of a script, to make which packages are required clear.</p>
<pre class="r"><code>library(ggplot2)</code></pre>
<p>once loaded you can use the functions in the package as we have done so far</p>
<p>We can use sessionInfo() to display information on R and the loaded packages.</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 19042)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
## [3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
## [5] LC_TIME=English_United Kingdom.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] docstring_1.0.0 forcats_0.5.1   stringr_1.4.0   purrr_0.3.4     readr_2.0.1     tidyr_1.1.3    
##  [7] tibble_3.1.3    tidyverse_1.3.1 dplyr_1.0.7     ggplot2_3.3.5  
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.1.1 xfun_0.26        bslib_0.3.0      haven_2.4.3      colorspace_2.0-2
##  [6] vctrs_0.3.8      generics_0.1.0   htmltools_0.5.2  yaml_2.2.1       utf8_1.2.2      
## [11] rlang_0.4.11     jquerylib_0.1.4  pillar_1.6.3     glue_1.4.2       withr_2.4.2     
## [16] DBI_1.1.1        dbplyr_2.1.1     modelr_0.1.8     readxl_1.3.1     lifecycle_1.0.1 
## [21] cellranger_1.1.0 munsell_0.5.0    gtable_0.3.0     rvest_1.0.1      evaluate_0.14   
## [26] knitr_1.34       tzdb_0.1.2       fastmap_1.1.0    fansi_0.5.0      highr_0.9       
## [31] broom_0.7.9      Rcpp_1.0.7       scales_1.1.1     backports_1.2.1  jsonlite_1.7.2  
## [36] fs_1.5.0         hms_1.1.1        digest_0.6.28    stringi_1.7.3    grid_4.1.0      
## [41] cli_3.0.1        tools_4.1.0      magrittr_2.0.1   sass_0.4.0       crayon_1.4.1    
## [46] pkgconfig_2.0.3  ellipsis_0.3.2   xml2_1.3.2       reprex_2.0.1     lubridate_1.7.10
## [51] roxygen2_7.1.2   rstudioapi_0.13  assertthat_0.2.1 rmarkdown_2.11   httr_1.4.2      
## [56] R6_2.5.1         compiler_4.1.0</code></pre>
<div id="documentation" class="section level2">
<h2>Documentation</h2>
<p>all packages have more or less extensive documentation. A quick Google search of your package and R or CRAN should bring up a documentation page similar to the one for <a href="https://cran.r-project.org/web/packages/cowplot/index.html">cowplot</a>. All packages have a reference manual, detailing one by one the use of all the functions. This is the same information that the help panel uses so it can be quite dry. Additionally most packages have a vignette, introducing the package and showing how to actually use it to solve a problem</p>
<p>we can also view the documentation from R with</p>
<pre class="r"><code>browseVignettes(&quot;tidyverse&quot;) # shows the vignette
?cowplot # shows the reference manual</code></pre>
</div>
<div id="citing-packages" class="section level2">
<h2>citing packages</h2>
<p>Citing the packages used in your work is important, for reproducibility and to acknowledge the work done by the package authors, in R more often than not, academics. You should always make an effort to cite the packages used in your analysis.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> We can use the <code>citation("package")</code> function to retrieve the citation information</p>
<p>for instance the IRanges package suggests you cite a PLOS Computational Biology</p>
<pre class="r"><code>citation(&quot;IRanges&quot;)</code></pre>
<pre><code>## 
##   Lawrence M, Huber W, Pag\`es H, Aboyoun P, Carlson M, et al. (2013) Software for
##   Computing and Annotating Genomic Ranges. PLoS Comput Biol 9(8): e1003118.
##   doi:10.1371/journal.pcbi.1003118
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {Software for Computing and Annotating Genomic Ranges},
##     author = {Michael Lawrence and Wolfgang Huber and Herv\&#39;e Pag\`es and Patrick Aboyoun and Marc Carlson and Robert Gentleman and Martin Morgan and Vincent Carey},
##     year = {2013},
##     journal = {{PLoS} Computational Biology},
##     volume = {9},
##     issue = {8},
##     doi = {10.1371/journal.pcbi.1003118},
##     url = {http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118},
##   }</code></pre>
</div>
</div>
<div id="key-packages" class="section level1">
<h1>key packages</h1>
<p>This section will introduce some key packages for data analysis.</p>
<div id="tidyverse" class="section level2">
<h2>Tidyverse</h2>
<p>the tidyverse is a collection of closely related packages designed that an be installed and loaded as one tidyverse They share a similar philosophy and can be very powerful for data manipulation. Note that the way of doing things in tidy R is often quite different from how you would operate in base R.</p>
<p>The tidyverse packages provide functions to clean and manipulate data. The concept of tidy data is the key concpet in all tidy packages. In tidy data:</p>
<ol style="list-style-type: decimal">
<li>Every column is a variable.</li>
<li>Every row is an observation.</li>
<li>Every cell is a single value.</li>
</ol>
<p>you can read more about tidy data <a href="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html">here</a></p>
<p>The tidyverse is composed of:</p>
<ul>
<li>ggplot2</li>
<li>tibble</li>
<li>tidyr</li>
<li>readr</li>
<li>purr</li>
<li>dyplr</li>
<li>stringr</li>
<li>forcats</li>
</ul>
</div>
<div id="ggplot2" class="section level2">
<h2>ggplot2</h2>
<p>ggplot2 is a plotting package. while base R plots can become very beautiful with lots of customisation, ggplot2 can easily produces publication quality graphics with comparatively little effort</p>
<p>GGplot is based on the grammar of graphics, a whole theory of how to describe a plot whose explanation is best left to the <a href="https://ggplot2-book.org/index.html">official book</a>. in short a ggplot2 graph is composed of 3 main elements: data, a geometry representing the individual data points and a mapping</p>
<p>for a gallery of what is possible with ggplot and other plotting packages see <a href="https://www.r-graph-gallery.com/histogram">r graph gallery</a> ggplot2’s grammar can appear complex at first, so we’ll cover some common plots in session 2.</p>
</div>
<div id="bioconductor" class="section level2">
<h2>Bioconductor</h2>
<p>Bioconductor is a repository of bioinformatics packages, which are not accessible from CRAN. We can install bioconductor following <a href="http://bioconductor.org/install/">their instructions</a>.</p>
<pre class="r"><code>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install()</code></pre>
<p>This will install a set of core packages, additional packages (i.e maftools) can be installed by specifying their name.</p>
<pre class="r"><code>BiocManager::install(&quot;maftools&quot;)</code></pre>
<p>We’ll use bioconductor and maftools in <a href="session3.Rmd">session 3</a></p>
</div>
<div id="r-markdown" class="section level2">
<h2>R markdown</h2>
<p>Compared to a script, which contains only your code and plain text comments, R markdown can be a very powerful way to present your R code, its output and comments in formatted text which can be shared in a wide range of formats including html pages, pdfs, word docs, and powerpoint presentations.</p>
<p>The base idea is that there are code chunks, which get executed with the output printed below and formatted text. Behind the scene, this website is automatically generated by a series of R notebooks, available on <a href="https://github.com/sdangelis/Bioinformatics-cafe-tutorials">github</a> but R Markdown can do much more! You can read all about R Markdown in this <a href="https://bookdown.org/yihui/rmarkdown/">book</a>.</p>
</div>
</div>
<div id="namespaces" class="section level1">
<h1>Namespaces</h1>
<p>Functions are conventionally named with a short, descriptive verb. Sometimes two packages use the same name for a function. In this case the most recently loaded package will overwrite the previous one</p>
<p>A very common case is the filter function in dplyr which overrides the default R function. Despite the shared name they do very different things.</p>
<p>The base R filter function applies a linear filter to a time series</p>
<pre class="r"><code>x &lt;- 1:100
filter(x, rep(1, 3))</code></pre>
<p>once we load dplyr the function filter filters rows in a data frame that satisfy a condition. In this case we can ask for hypothetical samples with less then 10 cells. Note: The variable cells is accessed directly rather than using <code>$cells</code>. This is one of the differences between normal and tidy R.</p>
<pre class="r"><code>library(dplyr, verbose = F)

sample=1:10
cells=1:10 * 4
experimnent &lt;- data.frame(sample, cells)
filter(experimnent, cells &lt; 10) </code></pre>
<pre><code>##   sample cells
## 1      1     4
## 2      2     8</code></pre>
<p>But what if we wanted the base R version? R has a very elegant built-in way to select which package we want to get our function from <code>namespace::function</code><a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. Generally speaking all packages have their own names pace so we can explicitly ask for the filter function in base R <a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> with <code>stats::filter()</code>.</p>
<pre class="r"><code>x &lt;- 1:100
stats::filter(x, rep(1, 3))</code></pre>
<pre><code>## Time Series:
## Start = 1 
## End = 100 
## Frequency = 1 
##   [1]  NA   6   9  12  15  18  21  24  27  30  33  36  39  42  45  48  51  54  57  60  63  66  69  72
##  [25]  75  78  81  84  87  90  93  96  99 102 105 108 111 114 117 120 123 126 129 132 135 138 141 144
##  [49] 147 150 153 156 159 162 165 168 171 174 177 180 183 186 189 192 195 198 201 204 207 210 213 216
##  [73] 219 222 225 228 231 234 237 240 243 246 249 252 255 258 261 264 267 270 273 276 279 282 285 288
##  [97] 291 294 297  NA</code></pre>
<p>Arguably, it can be worth to use the <code>::</code> operator when you use a function from a package to explicitly state what packages are required and isolate that code from whatever package we have loaded with library. This is suggested by the <a href="https://google.github.io/styleguide/Rguide.html">Google R Style Guide</a>.<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a></p>
</div>
<div id="packages-for-session-2" class="section level1">
<h1>Packages for session 2</h1>
<p>before the next session make sure you install the following packages - tidyverse - cowplot - Bioconductor</p>
<pre class="r"><code># you can use c() to create a vector of all the pacakges 
install.packages(c(&quot;tidyverse&quot;, &quot;ggplot2&quot;, &quot;cowplot&quot;))</code></pre>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>but wait? what is compilation?. unlike R some programming languages need to be translated from a human readable text format to a binary file understandable by the computer.R packages often interface with code in compiled languages as it is generally faster<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Of course, packages are often built on top of other packages, in most cases it would not be reasonable to cite all the packages used and all their dependencies.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Like C++ but unlike Python where the <code>import x as y</code> syntax is used to solve name conflicts<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>Built-in functions can belong to few different names paces,for instance base, stats, utils, graphics<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>This introduces an additional nuance, calling library() will load said package in the global environment.affecting all your code, using :: will only affect that function call.:: is the to go when writing a package<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
