<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>R packages and how to use them</title>

<script src="site_libs/header-attrs-2.8/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R tutorial</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">About</a>
</li>
<li>
  <a href="setup.html">setup</a>
</li>
<li>
  <a href="session1.html">session 1</a>
</li>
<li>
  <a href="packages.html">packages</a>
</li>
<li>
  <a href="session2.html">session 2</a>
</li>
<li>
  <a href="session3.html">session 3</a>
</li>
<li>
  <a href="reading.html">reading &amp; resources</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">R packages and how to use them</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#installing-packages">Installing packages</a>
<ul>
<li><a href="#packages-for-session-2">Packages for session 2</a></li>
</ul></li>
<li><a href="#using-packages">Using packages</a>
<ul>
<li><a href="#citing-packages">citing packages</a></li>
</ul></li>
<li><a href="#key-packages">key packages</a>
<ul>
<li><a href="#tidyverse">Tidyverse</a>
<ul>
<li><a href="#tidy-data">Tidy data</a></li>
</ul></li>
<li><a href="#ggplot2">ggplot2</a></li>
<li><a href="#bioconductor">Bioconductor</a></li>
<li><a href="#namespaces">namespaces</a></li>
<li><a href="#r-markdown">R markdown</a></li>
</ul></li>
</ul>
</div>

<p>So far we’ve only used functions supplied as part of base R. However the real power of R comes from the ability to import functions the community of R programmers has built. These come in packages, which are a series of related functions bundled together</p>
<p>generally speaking before implementing your function it is always worth checking if there is a well-maintained and documented package that performs what you would like to do.</p>
<div id="installing-packages" class="section level2">
<h2>Installing packages</h2>
<p>Most R packages belong to a repository known as CRAN. CRAN packages can be installed with a function</p>
<pre class="r"><code>install.packages(&quot;package_name&quot;)</code></pre>
<p>Often R will ask for your permission to install packages from source rather than precompiled binaries. If you want to use the most recent version in CRAN say yes, if you want to skip a potentially time consuming step say no. Installing packages from source on Windows PCs requires an additional component <a href="https://cran.r-project.org/bin/windows/Rtools/">RTools</a></p>
<p>But wait what does compiling mean? Unlike R and python where source code can be run directly, some programming languages need to be converted into human-unreadable machine code that is somewhat specific to a system before any code can be run. This step is known as compilation. In short this allows for optimisation of the resulting machine code at the cost of flexibility, as things like what type variables are can be decided in advance. R packages often interface with compiled code to speed up computation</p>
<div id="packages-for-session-2" class="section level3">
<h3>Packages for session 2</h3>
<p>before the next session try to install the following packages - tidyverse - cowplot - -</p>
<pre class="r"><code># you can use c() to create a vector of all the pacakges 
install.packages(c(&quot;tidyverse&quot;))</code></pre>
</div>
</div>
<div id="using-packages" class="section level2">
<h2>Using packages</h2>
<p>packages can be loaded by using the library function. Conventionally, packages are loaded at the beginning of a script, to make clear which packages the script needs to run</p>
<pre class="r"><code>library(cowplot)</code></pre>
<p>once loaded you can use the functions in the package as normal</p>
<p>Often it is useful to record which packages are loaded alongside their version, in this case we can use sessionInfo() to display information on R and the installed packages.</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 18363)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
## [3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
## [5] LC_TIME=English_United Kingdom.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] RColorBrewer_1.1-2 cowplot_1.1.1      broom_0.7.6       
## 
## loaded via a namespace (and not attached):
##  [1] highr_0.9         pillar_1.6.1      bslib_0.2.5.1     compiler_4.1.0    dbplyr_2.1.1      jquerylib_0.1.4  
##  [7] tools_4.1.0       digest_0.6.27     jsonlite_1.7.2    evaluate_0.14     lifecycle_1.0.0   tibble_3.1.2     
## [13] gtable_0.3.0      pkgconfig_2.0.3   rlang_0.4.11      DBI_1.1.1         cli_2.5.0         yaml_2.2.1       
## [19] xfun_0.23         dplyr_1.0.6       stringr_1.4.0     knitr_1.33        generics_0.1.0    vctrs_0.3.8      
## [25] sass_0.4.0        grid_4.1.0        tidyselect_1.1.1  glue_1.4.2        R6_2.5.0          fansi_0.5.0      
## [31] rmarkdown_2.8     tidyr_1.1.3       purrr_0.3.4       ggplot2_3.3.3     magrittr_2.0.1    backports_1.2.1  
## [37] scales_1.1.1      ellipsis_0.3.2    htmltools_0.5.1.1 assertthat_0.2.1  colorspace_2.0-1  utf8_1.2.1       
## [43] stringi_1.6.2     munsell_0.5.0     crayon_1.4.1</code></pre>
<p>In the rare event you’d need to unload a package you can use <code>detach("package:name", unload = TRUE)</code>. Be careful with this, it is often preferable to restart R form scratch. for instance we can now remove cowplot from the environment, and see it disappear from the session information</p>
<pre class="r"><code>detach(&quot;package:cowplot&quot;, unload = TRUE)
sessionInfo()</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 18363)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
## [3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
## [5] LC_TIME=English_United Kingdom.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] RColorBrewer_1.1-2 broom_0.7.6       
## 
## loaded via a namespace (and not attached):
##  [1] highr_0.9         pillar_1.6.1      bslib_0.2.5.1     compiler_4.1.0    dbplyr_2.1.1      jquerylib_0.1.4  
##  [7] tools_4.1.0       digest_0.6.27     jsonlite_1.7.2    evaluate_0.14     lifecycle_1.0.0   tibble_3.1.2     
## [13] gtable_0.3.0      pkgconfig_2.0.3   rlang_0.4.11      DBI_1.1.1         cli_2.5.0         yaml_2.2.1       
## [19] xfun_0.23         dplyr_1.0.6       stringr_1.4.0     knitr_1.33        generics_0.1.0    vctrs_0.3.8      
## [25] sass_0.4.0        grid_4.1.0        tidyselect_1.1.1  glue_1.4.2        R6_2.5.0          fansi_0.5.0      
## [31] rmarkdown_2.8     tidyr_1.1.3       purrr_0.3.4       ggplot2_3.3.3     magrittr_2.0.1    backports_1.2.1  
## [37] scales_1.1.1      ellipsis_0.3.2    htmltools_0.5.1.1 assertthat_0.2.1  colorspace_2.0-1  utf8_1.2.1       
## [43] stringi_1.6.2     munsell_0.5.0     crayon_1.4.1</code></pre>
<div id="citing-packages" class="section level3">
<h3>citing packages</h3>
<p>Citing the packages used in your work is important, both for reproducibility and to acknowledge the work done by the package authors. You should always make an effort to cite the packages used in your analysis. We can use the <code>citation("package")</code> function to retrieve the citation information the authors have added to their packages</p>
<p>for instance the IRanges package suggests you cite a PLOS Computational Biology</p>
<pre class="r"><code>citation(&quot;IRanges&quot;)</code></pre>
<pre><code>## 
##   Lawrence M, Huber W, Pag\`es H, Aboyoun P, Carlson M, et al. (2013) Software for Computing and Annotating
##   Genomic Ranges. PLoS Comput Biol 9(8): e1003118. doi:10.1371/journal.pcbi.1003118
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {Software for Computing and Annotating Genomic Ranges},
##     author = {Michael Lawrence and Wolfgang Huber and Herv\&#39;e Pag\`es and Patrick Aboyoun and Marc Carlson and Robert Gentleman and Martin Morgan and Vincent Carey},
##     year = {2013},
##     journal = {{PLoS} Computational Biology},
##     volume = {9},
##     issue = {8},
##     doi = {10.1371/journal.pcbi.1003118},
##     url = {http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118},
##   }</code></pre>
<p>Of course, packages are often built on top of other packages, in most conditions it would not be reasonable to cite all the packages used and all their dependencies.</p>
</div>
</div>
<div id="key-packages" class="section level1">
<h1>key packages</h1>
<p>This section will introduce some key packages for the analysis of biological data.</p>
<div id="tidyverse" class="section level2">
<h2>Tidyverse</h2>
<p>the tidyverse is a family of packages designed by the people behind RStudio. they can be installed and loaded as one tidyveres They share a similar philosophy and can be very powerful for data manipulation. However, it is important to note that the way of doing things in tidy R is often different form how you would operate on data in base R. It is important not to expect tidy R package to behave like base R. We’ll see some of these distinctions in session 2.</p>
<div id="tidy-data" class="section level3">
<h3>Tidy data</h3>
<p>the whole idea behind the tidyverse is that tidy data is easy to</p>
<p>The tidyverse packages provide methods to clean messy data and manipulate</p>
<ul>
<li></li>
<li></li>
<li></li>
<li></li>
<li></li>
<li></li>
<li></li>
<li></li>
</ul>
</div>
</div>
<div id="ggplot2" class="section level2">
<h2>ggplot2</h2>
<p>ggplot2 is a plotting package. while base R plots can become very beutiful with lots of customisation, ggplot2 can easily produces publication quality graphics.</p>
<p>GGplot is based on the grammar of graphics, a whole theory of how to describe a plot whose explanation is best left to the (official book)[<a href="https://ggplot2-book.org/index.html" class="uri">https://ggplot2-book.org/index.html</a>]. in short a ggplot2 graph is composed of at least 3 elements: data, a geometry representing the individual and a mapping</p>
<p>for a gallery of what is possible with ggplot and other plotting packages see <a href="https://www.r-graph-gallery.com/histogram">r graph gallery</a> ggplot2’s grammar can appear complex at first, so we’ll cover some common plots in session 2.</p>
</div>
<div id="bioconductor" class="section level2">
<h2>Bioconductor</h2>
<p>bioconductor is a repository of bioinformatics packages, which are not accessible We can install bioconductor following <a href="">their instructions</a> we can see here another example of the <code>namespace::function</code> syntax</p>
<pre class="r"><code>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install()</code></pre>
<p>This will install a set of core packages, additional packages (i.e AneuFinder) can be installed by specifying their name.</p>
<pre class="r"><code>BiocManager::install(&quot;AneuFinder&quot;)</code></pre>
</div>
<div id="namespaces" class="section level2">
<h2>namespaces</h2>
<p>Functions are conventionally named with a short, descriptive verb. However sometimes two packages use the same name for a function. Sometimes a the most reacently loaded package will overwirte the name</p>
<p>A very common case is the filter function in the dyplr package which overrides the default R function. Despite the shared name they do very differnt things</p>
<p>(base R) apply a linear filter to a timeseries</p>
<pre class="r"><code>x &lt;- 1:100
filter(x, rep(1, 3))</code></pre>
<pre><code>## Time Series:
## Start = 1 
## End = 100 
## Frequency = 1 
##   [1]  NA   6   9  12  15  18  21  24  27  30  33  36  39  42  45  48  51  54  57  60  63  66  69  72  75  78  81  84  87
##  [30]  90  93  96  99 102 105 108 111 114 117 120 123 126 129 132 135 138 141 144 147 150 153 156 159 162 165 168 171 174
##  [59] 177 180 183 186 189 192 195 198 201 204 207 210 213 216 219 222 225 228 231 234 237 240 243 246 249 252 255 258 261
##  [88] 264 267 270 273 276 279 282 285 288 291 294 297  NA</code></pre>
<p>once we load tidyr we see the function no longer works</p>
<pre class="r"><code>library(tidyr)
filter(x, rep(1, 3))</code></pre>
<pre><code>## Time Series:
## Start = 1 
## End = 100 
## Frequency = 1 
##   [1]  NA   6   9  12  15  18  21  24  27  30  33  36  39  42  45  48  51  54  57  60  63  66  69  72  75  78  81  84  87
##  [30]  90  93  96  99 102 105 108 111 114 117 120 123 126 129 132 135 138 141 144 147 150 153 156 159 162 165 168 171 174
##  [59] 177 180 183 186 189 192 195 198 201 204 207 210 213 216 219 222 225 228 231 234 237 240 243 246 249 252 255 258 261
##  [88] 264 267 270 273 276 279 282 285 288 291 294 297  NA</code></pre>
<p>R has a built-in way to select which package, or namespace, we want to get our function from <code>namespace::function</code></p>
<p>we can thus restore the default behavior by explicitly asking for the base filter, which is contained in the stats namespace.</p>
<pre class="r"><code>stats::filter(x, rep(1, 3))</code></pre>
<pre><code>## Time Series:
## Start = 1 
## End = 100 
## Frequency = 1 
##   [1]  NA   6   9  12  15  18  21  24  27  30  33  36  39  42  45  48  51  54  57  60  63  66  69  72  75  78  81  84  87
##  [30]  90  93  96  99 102 105 108 111 114 117 120 123 126 129 132 135 138 141 144 147 150 153 156 159 162 165 168 171 174
##  [59] 177 180 183 186 189 192 195 198 201 204 207 210 213 216 219 222 225 228 231 234 237 240 243 246 249 252 255 258 261
##  [88] 264 267 270 273 276 279 282 285 288 291 294 297  NA</code></pre>
</div>
<div id="r-markdown" class="section level2">
<h2>R markdown</h2>
<p>Compared to a script, which contains only your code and plain text comments, R markdown can be a very powerful way to present your R code, its output and comments in markdown formatted text which can be shared in a wide range of formats including pdfs, office docs, presentations.</p>
<p>To illustrate the power of r markdown this website is automatically generated by a series of R notebooks, which you can see <a href="https://github.com/sdangelis/Bioinformatics-cafe-tutorials">here</a></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
